### PR TITLE
Add `AccountRegistry` support for DAI `permit()` function

### PR DESCRIPTION
**Pull request by thomas-waite**
## Summary
This PR adds support to the `AccountRegistry.sol` for the DAI `permit()` function. 

The new DAI `permit()` function allows ERC20 spending approval to be granted via an offchain signature, rather than the sending of a transaction. 

Background on the `DAI.permit()` method: https://docs.makerdao.com/smart-contract-modules/dai-module/dai-detailed-documentation 

## Description
A new `AccountRegistry` behaviour contract, `Behaviour20200220.sol` has been added. This defines two methods with unique function signatures:

1. `deposit()` - convert DAI into zkDAI
2. `permit()`- call `DAI.permit()`

Given that `permit()` is currently not part of the ERC20 standard, and is a DAI specific method, it is importantly that the `AccountRegistry` supports both legacy approval flows and the new `permit()` approval flow. 

To this end, both the legacy `Behaviour20200106.deposit()` method and the new `Behaviour20200220.deposit()` method are available - they have different signatures as `Behaviour20200220.deposit()` takes two additional parameters `signature` and `nonce`. 

`signature` is an EIP712 signature that is passed to the `DAI.permit()` method. It is used to approve the `proxyContract` to spend DAI on behalf of the user, rather than having to send a `dai.approve()` transaction.  

`nonce` is a record of the number of `permit()` calls have been made and is used for signature replay protection in the DAI contract. 

### `Behaviour20200220.permit()`
This method has been given the visibility modifier of `public` so that users are able to directly call `permit()` and so revoke approval at a later date if they wish/need to.

### CODE REVIEW

File: packages/extension/src/ui/apis/asset/permitERC20Allowance.js [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+            .useContract('ERC20')
+            .at(linkedTokenAddress)
+            .method('nonces')
+            .send(
+                allowanceSpender,
+            );
+    } catch (e) {
+        error = e;
+    }
+    const exipry = Date.now() + 24 * 60 * 60;
```
Comment (ArnSch) [inline packages/extension/src/ui/apis/asset/permitERC20Allowance.js]:
```suggestion
    const expiry = Date.now() + 24 * 60 * 60;
```

(this is going to go anyways)

File: packages/protocol/contracts/AccountRegistry/epochs/20200220/Behaviour20200220.sol [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+    * 
+    * @param _registryOwner - owner of the zkAsset
+    * @param _owner - owner of the ERC20s being deposited
+    * @param _proofHash - hash of the zero-knowledge deposit proof
+    * @param _proofData - cryptographic data associated with the zero-knowledge proof
+    * @param _value - number of ERC20s being deposited
+    * @param signature - EIP712 signature supplied to permit() for granting approval
+    * @param nonce - permit nonce used for signature replay protection
+     */
+    function deposit(
```
Comment (ArnSch) [inline packages/protocol/contracts/AccountRegistry/epochs/20200220/Behaviour20200220.sol:51]: As discussed, should make the permit call to transfer tokens to this contract first, then rely on `super.deposit()` to call existing deposit function in previous inherited behaviour to make changes explicit, and increase confidence that no changes in logic have occurred beyond expected ones.

**Review by ArnSch — CHANGES_REQUESTED** [review]
Looks good. We discussed changes to deposit function, and to make the permit timestamp much more constrained, but otherwise good to go.

File: packages/protocol/contracts/AccountRegistry/epochs/20200220/Behaviour20200220.sol (hunk 2 above)
Comment (thomas-waite) [reply packages/protocol/contracts/AccountRegistry/epochs/20200220/Behaviour20200220.sol:51]: Yeah good idea, have done

**Comment by ArnSch** [comment]
@thomas-waite still one lint error (`expiry` declared but not used in `depositWithPermit`), but did not know whether that was WIP so left as is. Otherwise fixed lint issues.

**Review by ArnSch — APPROVED** [review]

**Comment by joeandrews** [comment]
:tada: This PR is included in version 0.17.0 :tada:

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:

**Comment by joeandrews** [comment]
:tada: This PR is included in version 1.18.0 :tada:

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:

**Comment by joeandrews** [comment]
:tada: This PR is included in version 1.20.0 :tada:

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:

**Comment by joeandrews** [comment]
:tada: This PR is included in version 2.4.0 :tada:

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:

**Comment by joeandrews** [comment]
:tada: This PR is included in version 1.6.0 :tada:

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:

**Comment by joeandrews** [comment]
:tada: This PR is included in version 1.3.2 :tada:

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:

**Comment by joeandrews** [comment]
:tada: This PR is included in version 1.1.4 :tada:

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:

**Comment by joeandrews** [comment]
:tada: This PR is included in version 1.9.0 :tada:

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:

**Pull request merged**